File: agent/harmonia_agent/operator_instructions.py
"""Strict cross-runtime contract for operator instruction and clarification provenance."""
from __future__ import annotations

import hashlib
import json
from typing import Literal

from pydantic import BaseModel, ConfigDict, Field, model_validator


class StrictModel(BaseModel):
    model_config = ConfigDict(extra="forbid")


class InstructionTurnProvenance(StrictModel):
    turnId: str = Field(pattern=r"^[A-Za-z0-9_-]{1,128}$")
    messageDigest: str = Field(pattern=r"^[a-f0-9]{64}$")
    resolvedField: Literal[
        "expectedOutcome", "target", "rights", "requestedOutputs",
        "sources", "strategyContext", "activeStrategy",
    ] | None = None


class OperatorInstructionContext(StrictModel):
    originalOperatorBrief: str = Field(min_length=1, max_length=20000)
    resolvedInstructions: str = Field(min_length=1, max_length=20000)
    intakeDraftId: str = Field(pattern=r"^[a-f0-9]{64}$")
    intakeRevision: int = Field(gt=0, le=9007199254740991)
    answerTurnIds: list[str] = Field(max_length=49)
    turnProvenance: list[InstructionTurnProvenance] = Field(min_length=1, max_length=50)
    contextDigest: str = Field(pattern=r"^[a-f0-9]{64}$")

    @model_validator(mode="after")
    def verify_provenance(self):
        turn_ids = [turn.turnId for turn in self.turnProvenance]
        if len(turn_ids) != len(set(turn_ids)):
            raise ValueError("instruction turn identities must be unique")
        if self.answerTurnIds != turn_ids[1:]:
            raise ValueError("answer turn identities must match clarification provenance")
        unsigned = self.model_dump(exclude={"contextDigest"}, mode="json", exclude_none=True)
        encoded = json.dumps(unsigned, sort_keys=True, separators=(",", ":"), ensure_ascii=False).encode()
        if hashlib.sha256(encoded).hexdigest() != self.contextDigest:
            raise ValueError("instruction context digest mismatch")
        return self


def validate_operator_instruction_context(value: object) -> dict:
    return OperatorInstructionContext.model_validate(value).model_dump(mode="json", exclude_none=True)


def validate_provider_instruction_binding(sealed_request: dict) -> dict:
    context = validate_operator_instruction_context(sealed_request.get("instructionContext"))
    request = sealed_request.get("request")
    if not isinstance(request, dict) or request.get("prompt") != context["resolvedInstructions"]:
        raise ValueError("provider prompt differs from sealed operator instructions")
    return context

File: src/lib/operatorInstructions.ts
import { createHash } from "node:crypto";
import { z } from "zod";

const digestSchema = z.string().regex(/^[a-f0-9]{64}$/);
const turnIdSchema = z.string().regex(/^[A-Za-z0-9_-]{1,128}$/);
const resolvedFieldSchema = z.enum(["expectedOutcome", "target", "rights", "requestedOutputs", "sources", "strategyContext", "activeStrategy", "appendConstraints"]);

const canonical = (value: unknown): string => {
  if (value === null || typeof value === "boolean" || typeof value === "string") return JSON.stringify(value);
  if (typeof value === "number") {
    if (!Number.isFinite(value)) throw new Error("non-finite instruction context value");
    return JSON.stringify(value);
  }
  if (Array.isArray(value)) return `[${value.map(canonical).join(",")}]`;
  if (typeof value === "object") return `{${Object.entries(value as Record<string, unknown>)
    .filter(([, item]) => item !== undefined)
    .sort(([a], [b]) => a.localeCompare(b))
    .map(([key, item]) => `${JSON.stringify(key)}:${canonical(item)}`).join(",")}}`;
  throw new Error("unsupported instruction context value");
};
const sha = (value: unknown) => createHash("sha256").update(canonical(value)).digest("hex");

export const instructionTurnProvenanceSchema = z.object({
  turnId: turnIdSchema,
  messageDigest: digestSchema,
  resolvedField: resolvedFieldSchema.optional(),
}).strict();

const unsignedInstructionContextSchema = z.object({
  originalOperatorBrief: z.string().min(1).max(20000),
  resolvedInstructions: z.string().min(1).max(20000),
  intakeDraftId: digestSchema,
  intakeRevision: z.number().int().positive().max(Number.MAX_SAFE_INTEGER),
  answerTurnIds: z.array(turnIdSchema).max(49),
  turnProvenance: z.array(instructionTurnProvenanceSchema).min(1).max(50),
}).strict().superRefine((value, context) => {
  const turnIds = value.turnProvenance.map(turn => turn.turnId);
  if (new Set(turnIds).size !== turnIds.length) context.addIssue({ code: "custom", path: ["turnProvenance"], message: "instruction turn identities must be unique" });
  if (canonical(value.answerTurnIds) !== canonical(turnIds.slice(1))) context.addIssue({ code: "custom", path: ["answerTurnIds"], message: "answer turn identities must match clarification provenance" });
});

export const operatorInstructionContextSchema = unsignedInstructionContextSchema.extend({
  contextDigest: digestSchema,
}).strict().superRefine((value, context) => {
  const { contextDigest, ...unsigned } = value;
  if (contextDigest !== sha(unsigned)) context.addIssue({ code: "custom", path: ["contextDigest"], message: "instruction context digest mismatch" });
});
export type OperatorInstructionContext = z.infer<typeof operatorInstructionContextSchema>;

export function sealOperatorInstructionContext(input: {
  draftId: string;
  revision: number;
  originalOperatorBrief: string;
  answers: Array<{ requestId: string; message: string; resolvedField?: z.infer<typeof resolvedFieldSchema> | null }>;
}): OperatorInstructionContext {
  const answers = input.answers.map(answer => ({
    turnId: answer.requestId,
    message: z.string().min(1).max(20000).parse(answer.message),
    messageDigest: sha(answer.message),
    ...(answer.resolvedField ? { resolvedField: answer.resolvedField } : {}),
  }));
  if (!answers.length || answers[0].message !== input.originalOperatorBrief) throw new Error("original operator brief must match the first intake turn");
  const unsigned = unsignedInstructionContextSchema.parse({
    originalOperatorBrief: input.originalOperatorBrief,
    resolvedInstructions: answers.map(answer => answer.message).join("\n\n"),
    intakeDraftId: input.draftId,
    intakeRevision: input.revision,
    answerTurnIds: answers.slice(1).map(answer => answer.turnId),
    turnProvenance: answers.map(({ turnId, messageDigest, resolvedField }) => ({ turnId, messageDigest, ...(resolvedField ? { resolvedField } : {}) })),
  });
  return operatorInstructionContextSchema.parse({ ...unsigned, contextDigest: sha(unsigned) });
}
